docs/profiler.rst

```diff
@@ -2,45 +2,238 @@
 Profiler
 ========
 
-Taichi's profiler can help you analyze the run-time cost of your program. There are two profiling systems in Taichi: ``ScopedProfiler`` and ``KernelProfiler``.
+Taichi's profiler can help you analyze the run-time cost of your program, i.e. performance.
+
+Currently there are three profiling systems in Taichi:
+
+- ``PythonProfiler``
+- ``KernelProfiler``
+- ``ScopedProfiler`` (for developers)
+
 
-ScopedProfiler
-##############
 
-1. ``ScopedProfiler`` measures time spent on the **host tasks** hierarchically.
+PythonProfiler
+--------------
 
-2. This profiler is automatically on. To show its results, call ``ti.print_profile_info()``. For example:
+``PythonProfiler`` basically measures time spent between ``start()`` and ``stop()`` using
+the Python-builtin function ``time.time()``.
+
+Profiling APIs
+**************
+
+There are 3 ways to use this profiler:
+
+1. ``ti.profiler.start()`` and ``ti.profiler.stop()``, these are the most fundemental APIs:
 
 .. code-block:: python
 
+    from time import sleep
     import taichi as ti
+    ti.init()
 
-    ti.init(arch=ti.cpu)
-    var = ti.var(ti.f32, shape=1)
+    def do_something_A():
+        sleep(0.01)
+
+    def do_something_B():
+        sleep(0.1)
+
+    ti.profiler.start('A')
+    do_something_A()
+    ti.profiler.stop('A')
+
+    ti.profiler.start('B')
+    do_something_B()
+    ti.profiler.stop('B')
+
+    ti.profiler.print()
+
+
+2. ``ti.profiler()``, this one makes our code cleaner:
+
+.. code-block:: python
+
+    from time import sleep
+    import taichi as ti
+    ti.init()
+
+    def do_something_A():
+        sleep(0.01)
+
+    def do_something_B():
+        sleep(0.1)
+
+    ti.profiler('A')  # start('A')
+    do_something_A()
+
+    ti.profiler('B')  # stop('A') & start('B')
+    do_something_B()
+    ti.profiler()     # stop('B')
+
+    ti.profiler.print()
+
+
+.. note::
+
+    Running 1 and 2 should obtain something like:
+
+    .. code-block:: none
+
+          min   |   avg   |   max   |  num  |  total  |    name
+         0.100s |  0.100s |  0.100s |    1x |  0.100s | B
+        10.10ms | 10.10ms | 10.10ms |    1x | 10.10ms | A
+
+
+3. ``@ti.profiler.timed``, this one is very intuitive when profiling kernels:
```

Comment:
Do we need to explain this decorator more? It is still a little confused maybe.

Reply:
:+1: Will do this once later this day.

Reply:
New version done! Please ff2 leave a review.